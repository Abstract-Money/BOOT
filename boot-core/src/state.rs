use std::collections::HashMap;

use cosmwasm_std::Addr;

use crate::BootError;

pub trait ChainState {
    type Out: StateInterface;
    fn state(&self) -> Self::Out;
}

/// Interface for setting the local (jsonified) state of the contracts on a chain
pub trait StateInterface: Clone {
    fn get_address(&self, contract_id: &str) -> Result<Addr, BootError>;
    fn set_address(&mut self, contract_id: &str, address: &Addr);
    fn get_code_id(&self, contract_id: &str) -> Result<u64, BootError>;
    fn set_code_id(&mut self, contract_id: &str, code_id: u64);
    fn get_checksum(&self, contract_id: &str) -> Result<String, BootError>;
    fn set_checksum(&mut self, contract_id: &str, checksum: &str);
    fn get_all_addresses(&self) -> Result<HashMap<String, Addr>, BootError>;
    fn get_all_code_ids(&self) -> Result<HashMap<String, u64>, BootError>;
}

// pub struct DaemonState{
//     pub path: String,
// }

// impl StateInterface for DaemonState {
//     fn address(&self, key: &str) -> String {
//         todo!()
//     }

//     fn save_address(&self,contract_id: &str, address: &str) {
//         todo!()
//     }
// }
