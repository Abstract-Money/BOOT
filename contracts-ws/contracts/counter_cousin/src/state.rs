use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::ContractError;
use cosmwasm_std::{Addr, Deps, MessageInfo};
use cw_storage_plus::Item;

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
pub struct State {
    pub count: i32,
    pub owner: Addr,
}

pub const STATE: Item<State> = Item::new("state");

pub fn assert_owner(deps: Deps, info: &MessageInfo) -> Result<(), ContractError> {
    let state = STATE.load(deps.storage)?;
    if info.sender != state.owner {
        return Err(ContractError::Unauthorized {});
    }
    Ok(())
}
