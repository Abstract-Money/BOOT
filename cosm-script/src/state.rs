use cosmwasm_std::Addr;

use crate::CosmScriptError;

pub trait ChainState {
    type Out: StateInterface;
    fn state(&self) -> Self::Out;
}

pub trait StateInterface {
    fn get_address(&self, contract_id: &str) -> Result<Addr, CosmScriptError>;
    fn set_address(&mut self, contract_id: &str, address: &Addr);
    fn get_code_id(&self, contract_id: &str) -> Result<u64, CosmScriptError>;
    fn set_code_id(&mut self, contract_id: &str, code_id: u64);
}

// pub struct DaemonState{
//     pub path: String,
// }

// impl StateInterface for DaemonState {
//     fn address(&self, key: &str) -> String {
//         todo!()
//     }

//     fn save_address(&self,contract_id: &str, address: &str) {
//         todo!()
//     }
// }
