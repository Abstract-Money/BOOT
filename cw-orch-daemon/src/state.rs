use super::error::DaemonError;
use crate::{chain_info::ChainKind, channel::GrpcChannel};

use ibc_chain_registry::chain::ChainData;
use serde::Serialize;
use serde_json::{json, Value};
use std::{env, fs::File, path::Path};
use tonic::transport::Channel;

/// Stores the chain information and deployment state.
/// Uses a simple JSON file to store the deployment information locally.
#[derive(Clone, Debug)]
pub struct DaemonState {
    /// this is passed via env var STATE_FILE
    pub json_file_path: String,
    /// Deployment identifier
    pub deployment_id: String,
    /// gRPC channel
    pub grpc_channel: Channel,
    /// Information about the chain
    pub chain_data: ChainData,
}

impl DaemonState {
    /// Creates a new state from the given chain data and deployment id.
    /// Attempts to connect to any of the provided gRPC endpoints.
    pub async fn new(
        mut chain_data: ChainData,
        deployment_id: String,
    ) -> Result<DaemonState, DaemonError> {
        if chain_data.apis.grpc.is_empty() {
            return Err(DaemonError::GRPCListIsEmpty);
        }

        log::info!("Found {} gRPC endpoints", chain_data.apis.grpc.len());

        // find working grpc channel
        let grpc_channel =
            GrpcChannel::connect(&chain_data.apis.grpc, &chain_data.chain_id).await?;

        // check if STATE_FILE en var is configured, default to state.json
        let mut json_file_path = env::var("STATE_FILE").unwrap_or("./state.json".to_string());

        // if the network we are connecting is a local kind, add it to the fn
        if chain_data.network_type == ChainKind::Local.to_string() {
            let name = Path::new(&json_file_path)
                .file_stem()
                .unwrap()
                .to_str()
                .unwrap();
            let folder = Path::new(&json_file_path)
                .parent()
                .unwrap()
                .to_str()
                .unwrap();

            json_file_path = format!("{folder}/{name}_local.json");
        }

        // Try to get the standard fee token (probably shortest denom)
        let shortest_denom_token = chain_data.fees.fee_tokens.iter().fold(
            chain_data.fees.fee_tokens[0].clone(),
            |acc, item| {
                if item.denom.len() < acc.denom.len() {
                    item.clone()
                } else {
                    acc
                }
            },
        );
        // set a single fee token
        chain_data.fees.fee_tokens = vec![shortest_denom_token];

        // build daemon state
        let state = DaemonState {
            json_file_path,
            deployment_id,
            grpc_channel,
            chain_data,
        };

        log::info!(
            "Writing daemon state JSON file: {:#?}",
            state.json_file_path
        );

        // write json state file
        crate::json_file::write(
            &state.json_file_path,
            &state.chain_data.chain_id.to_string(),
            &state.chain_data.chain_name,
            &state.deployment_id,
        );

        // finish
        Ok(state)
    }

    /// Get the state filepath and read it as json
    fn read_state(&self) -> serde_json::Value {
        crate::json_file::read(&self.json_file_path)
    }

    /// Retrieve a stateful value using the chainId and networkId
    pub fn get(&self, key: &str) -> Value {
        let json = self.read_state();
        json[&self.chain_data.chain_name][&self.chain_data.chain_id.to_string()][key].clone()
    }

    /// Set a stateful value using the chainId and networkId
    pub fn set<T: Serialize>(&self, key: &str, contract_id: &str, value: T) {
        let mut json = self.read_state();

        json[&self.chain_data.chain_name][&self.chain_data.chain_id.to_string()][key]
            [contract_id] = json!(value);

        serde_json::to_writer_pretty(File::create(&self.json_file_path).unwrap(), &json).unwrap();
    }
}
