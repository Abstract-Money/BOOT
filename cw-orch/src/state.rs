use crate::error::CwOrcError;
use cosmwasm_std::Addr;
use std::collections::HashMap;

/// State accessor trait
/// Indicates that the type has access to an underlying state
pub trait ChainState {
    type Out: StateInterface;
    fn state(&self) -> Self::Out;
}

/// Interface for setting the local state of the contracts on a chain
pub trait StateInterface: Clone {
    fn get_address(&self, contract_id: &str) -> Result<Addr, CwOrcError>;
    fn set_address(&mut self, contract_id: &str, address: &Addr);
    fn get_code_id(&self, contract_id: &str) -> Result<u64, CwOrcError>;
    fn set_code_id(&mut self, contract_id: &str, code_id: u64);
    fn get_all_addresses(&self) -> Result<HashMap<String, Addr>, CwOrcError>;
    fn get_all_code_ids(&self) -> Result<HashMap<String, u64>, CwOrcError>;
}
