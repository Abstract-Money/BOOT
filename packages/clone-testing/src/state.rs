use cosmwasm_std::Addr;
use cw_orch_core::{
    environment::{ChainInfoOwned, StateInterface},
    CwEnvError,
};
use cw_orch_daemon::DaemonState;
use itertools::Itertools;
use std::{collections::HashMap, sync::Arc};

#[derive(Clone, Debug)]
/// Mock state for testing, stores addresses and code-ids.
pub struct MockState {
    /// Deployed contract code ids
    pub code_ids: HashMap<String, u64>,
    /// Deployed contract addresses
    pub addresses: HashMap<String, Addr>,
    /// State read from file. Used to actually integrate with actual deployments
    pub daemon_state: DaemonState,
}

impl MockState {
    /// Creates a new empty mock state
    pub fn new(chain: ChainInfoOwned, deployment_id: &str) -> Self {
        Self {
            addresses: HashMap::new(),
            code_ids: HashMap::new(),
            daemon_state: DaemonState::new(
                DaemonState::state_file_path().unwrap(),
                &Arc::new(chain),
                deployment_id.to_string(),
                true,
                false,
            )
            .unwrap(),
        }
    }
}

impl StateInterface for MockState {
    fn get_address(&self, contract_id: &str) -> Result<Addr, CwEnvError> {
        // First we look for the address inside the mock state
        self.addresses
            .get(contract_id)
            .ok_or_else(|| CwEnvError::AddrNotInStore(contract_id.to_owned()))
            .map(|val| val.to_owned())
            // If not present, we look for it in the daemon state
            .or_else(|_| self.daemon_state.get_address(contract_id))
    }

    fn set_address(&mut self, contract_id: &str, address: &Addr) {
        self.addresses
            .insert(contract_id.to_string(), address.to_owned());
    }

    fn remove_address(&mut self, contract_id: &str) {
        self.addresses.remove(contract_id);
    }

    /// Get the locally-saved version of the contract's version on this network
    fn get_code_id(&self, contract_id: &str) -> Result<u64, CwEnvError> {
        self.code_ids
            .get(contract_id)
            .ok_or_else(|| CwEnvError::CodeIdNotInStore(contract_id.to_owned()))
            .map(|val| val.to_owned())
            // If not present, we look for it in the daemon state
            .or_else(|_| self.daemon_state.get_code_id(contract_id))
    }

    /// Set the locally-saved version of the contract's latest version on this network
    fn set_code_id(&mut self, contract_id: &str, code_id: u64) {
        self.code_ids.insert(contract_id.to_string(), code_id);
    }

    fn remove_code_id(&mut self, contract_id: &str) {
        self.code_ids.remove(contract_id);
    }

    fn get_all_addresses(&self) -> Result<HashMap<String, Addr>, CwEnvError> {
        let mock_addresses = self.addresses.clone();
        let daemon_addresses = self.daemon_state.get_all_addresses().unwrap_or_default();

        Ok(mock_addresses
            .into_iter()
            .chain(daemon_addresses)
            .unique()
            .collect())
    }

    fn get_all_code_ids(&self) -> Result<HashMap<String, u64>, CwEnvError> {
        let mock_code_ids = self.code_ids.clone();
        let daemon_code_ids = self.daemon_state.get_all_code_ids().unwrap_or_default();

        Ok(mock_code_ids
            .into_iter()
            .chain(daemon_code_ids)
            .unique()
            .collect())
    }
}

#[cfg(test)]
mod test {
    use cosmwasm_std::Addr;
    use cw_orch_core::{environment::StateInterface, CwEnvError};
    use cw_orch_daemon::networks::JUNO_1;
    use speculoos::prelude::*;

    use super::MockState;

    const CONTRACT_ID: &str = "123";
    const CONTRACT_ADDR: &str = "cosmos123";
    #[test]
    fn mock_state() {
        let mut mock = MockState::new(JUNO_1.into(), "default-id");
        env_logger::init();

        let unchecked_address = &Addr::unchecked(CONTRACT_ADDR);
        let code_id = 123u64;

        mock.set_address(CONTRACT_ID, unchecked_address);
        mock.set_code_id(CONTRACT_ID, code_id);

        // assert we get the right address
        let addr = mock.get_address(CONTRACT_ID).unwrap();
        asserting!(&"address is correct for contract_id")
            .that(unchecked_address)
            .is_equal_to(&addr);

        // assert we get the right code_id
        let fetched_id = mock.get_code_id(CONTRACT_ID).unwrap();
        asserting!(&"code_id is correct for contract_id")
            .that(&fetched_id)
            .is_equal_to(code_id);

        // assert we get AddrNotInStore error
        let missing_id = &"456";
        let error = mock.get_address(missing_id).unwrap_err();
        let error_msg = CwEnvError::AddrNotInStore(String::from(*missing_id)).to_string();
        asserting!(&(format!("Asserting we get CwEnvError: {}", error_msg)))
            .that(&error.to_string())
            .is_equal_to(CwEnvError::AddrNotInStore(String::from(*missing_id)).to_string());

        // assert we get CodeIdNotInStore error
        let error_msg = CwEnvError::CodeIdNotInStore(String::from(*missing_id)).to_string();
        let error = mock.get_code_id(missing_id).unwrap_err();
        asserting!(&(format!("Asserting we get CwEnvError: {}", error_msg)))
            .that(&error.to_string())
            .is_equal_to(CwEnvError::CodeIdNotInStore(String::from(*missing_id)).to_string());

        // validate we can get all addresses
        let total = mock.get_all_addresses().unwrap().len();
        asserting!(&"total addresses is one")
            .that(&total)
            .is_equal_to(1);

        // validate we can get all code_ids
        let total = mock.get_all_code_ids().unwrap().len();
        asserting!(&"total code_ids is one")
            .that(&total)
            .is_greater_than_or_equal_to(1)
    }
}
