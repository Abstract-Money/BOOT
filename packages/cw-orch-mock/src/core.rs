use std::{cell::RefCell, fmt::Debug, rc::Rc};

use cosmwasm_std::{
    testing::{MockApi, MockStorage},
    Addr, Coin, Empty, Event, Uint128,
};
use cw_multi_test::{
    ibc::IbcSimpleModule, App, AppBuilder, AppResponse, BankKeeper, Contract, DistributionKeeper,
    Executor, FailingModule, GovFailingModule, StakeKeeper, StargateFailing, WasmKeeper,
};
use cw_utils::NativeBalance;
use serde::Serialize;

use cw_orch_core::{
    contract::interface_traits::Uploadable,
    environment::TxHandler,
    environment::{
        queriers::bank::{BankQuerier},
        BankSetter, ChainState, IndexResponse, StateInterface,
    },
    CwEnvError,
};

use super::state::MockState;

pub type MockApp = App<
    BankKeeper,
    MockApi,
    MockStorage,
    FailingModule<Empty, Empty, Empty>,
    WasmKeeper<Empty, Empty>,
    StakeKeeper,
    DistributionKeeper,
    IbcSimpleModule,
    GovFailingModule,
    StargateFailing,
>;

/// Wrapper around a cw-multi-test [`App`](cw_multi_test::App) backend.
///
/// Stores a local state with a mapping of contract_id -> code_id/address
///
/// The state is customizable by implementing the [`StateInterface`] trait on a custom struct and providing it on the custom constructor.
///
/// The addresses used inside this environment are bech32 addresses. For instance, when creating a mock environment
/// let chain = Mock::new("sender");
/// the actual sender address can be generated using
/// let sender_addr = chain.addr_make("sender")
///
/// ## Example
/// ```
/// # use cosmwasm_std::{Addr, coin, Uint128};
/// use cw_orch_mock::Mock;
/// use cw_orch_core::environment::TxHandler;
///
/// let mock: Mock = Mock::new("sender");
///
/// // set a balance
/// mock.set_balance("sender", vec![coin(100u128, "token")]).unwrap();
///
/// // query the balance
/// let balance: Uint128 = mock.query_balance("sender", "token").unwrap();
/// assert_eq!(balance.u128(), 100u128);
/// ```
///
/// ## Example with custom state
/// ```
/// # use cosmwasm_std::{Addr, coin, Uint128};
/// use cw_orch_mock::Mock;
/// use cw_orch_core::environment::StateInterface;
/// // We just use the MockState as an example here, but you can implement your own state struct.
/// use cw_orch_mock::MockState as CustomState;
///
/// let mock: Mock = Mock::new_custom("sender", CustomState::new());
/// ```
#[derive(Clone)]
pub struct Mock<S: StateInterface = MockState> {
    /// Address used for the operations.
    pub sender: Addr,
    /// Inner mutable state storage for contract addresses and code-ids
    pub state: Rc<RefCell<S>>,
    /// Inner mutable cw-multi-test app backend
    pub app: Rc<RefCell<MockApp>>,
}

impl<S: StateInterface> Mock<S> {
    /// Set the bank balance of an address.
    pub fn set_balance(
        &self,
        address: impl Into<String>,
        amount: Vec<cosmwasm_std::Coin>,
    ) -> Result<(), CwEnvError> {
        self.app
            .borrow_mut()
            .init_modules(|router, _, storage| {
                router
                    .bank
                    .init_balance(storage, &Addr::unchecked(address.into()), amount)
            })
            .map_err(Into::into)
    }

    /// Adds the bank balance of an address.
    pub fn add_balance(
        &self,
        address: impl Into<String>,
        amount: Vec<cosmwasm_std::Coin>,
    ) -> Result<(), CwEnvError> {
        let addr = &Addr::unchecked(address.into());
        let b = self.query_all_balances(addr.clone())?;
        let new_amount = NativeBalance(b) + NativeBalance(amount);
        self.app
            .borrow_mut()
            .init_modules(|router, _, storage| {
                router
                    .bank
                    .init_balance(storage, addr, new_amount.into_vec())
            })
            .map_err(Into::into)
    }

    /// Set the balance for multiple coins at once.
    pub fn set_balances(
        &self,
        balances: &[(impl Into<String> + Clone, &[cosmwasm_std::Coin])],
    ) -> Result<(), CwEnvError> {
        self.app
            .borrow_mut()
            .init_modules(|router, _, storage| -> Result<(), CwEnvError> {
                for (addr, coins) in balances {
                    router.bank.init_balance(
                        storage,
                        &Addr::unchecked(addr.clone()),
                        coins.to_vec(),
                    )?;
                }
                Ok(())
            })
    }

    /// Query the (bank) balance of a native token for and address.
    /// Returns the amount of the native token.
    pub fn query_balance(
        &self,
        address: impl Into<String>,
        denom: &str,
    ) -> Result<Uint128, CwEnvError> {
        Ok(self
            .bank_querier()
            .balance(address, Some(denom.to_string()))?
            .first()
            .map(|c| c.amount)
            .unwrap_or_default())
    }

    /// Fetch all the balances of an address.
    pub fn query_all_balances(
        &self,
        address: impl Into<String>,
    ) -> Result<Vec<cosmwasm_std::Coin>, CwEnvError> {
        self.bank_querier().balance(address, None)
    }
}

impl Mock<MockState> {
    /// Create a mock environment with the default mock state.
    pub fn new(sender: impl Into<String>) -> Self {
        Mock::new_custom(sender, MockState::new())
    }

    pub fn new_with_chain_id(sender: impl Into<String>, chain_id: &str) -> Self {
        let chain = Mock::new_custom(sender, MockState::new());
        chain
            .app
            .borrow_mut()
            .update_block(|b| b.chain_id = chain_id.to_string());

        chain
    }

    pub fn with_chain_id(&mut self, chain_id: &str) {
        self.state.borrow_mut().set_chain_id(chain_id);
        self.app
            .borrow_mut()
            .update_block(|b| b.chain_id = chain_id.to_string());
    }
}

impl<S: StateInterface> Mock<S> {
    /// Create a mock environment with a custom mock state.
    /// The state is customizable by implementing the `StateInterface` trait on a custom struct and providing it on the custom constructor.
    pub fn new_custom(sender: impl Into<String>, custom_state: S) -> Self {
        let state = Rc::new(RefCell::new(custom_state));
        let app = Rc::new(RefCell::new(AppBuilder::new_custom().build(|_, _, _| {})));

        Self {
            sender: Addr::unchecked(sender),
            state,
            app,
        }
    }

    /// Upload a custom contract wrapper.
    /// Support for this is limited.
    pub fn upload_custom(
        &self,
        contract_id: &str,
        wrapper: Box<dyn Contract<Empty, Empty>>,
    ) -> Result<AppResponse, CwEnvError> {
        let code_id = self.app.borrow_mut().store_code(wrapper);
        // add contract code_id to events manually
        let mut event = Event::new("store_code");
        event = event.add_attribute("code_id", code_id.to_string());
        let resp = AppResponse {
            events: vec![event],
            ..Default::default()
        };
        let code_id = IndexResponse::uploaded_code_id(&resp)?;
        self.state.borrow_mut().set_code_id(contract_id, code_id);
        Ok(resp)
    }
}

impl<S: StateInterface> ChainState for Mock<S> {
    type Out = Rc<RefCell<S>>;

    fn state(&self) -> Self::Out {
        self.state.clone()
    }
}

// Execute on the test chain, returns test response type
impl<S: StateInterface> TxHandler for Mock<S> {
    type Response = AppResponse;
    type Error = CwEnvError;
    type ContractSource = Box<dyn Contract<Empty, Empty>>;
    type Sender = Addr;

    fn sender(&self) -> Addr {
        self.sender.clone()
    }

    fn set_sender(&mut self, sender: Self::Sender) {
        self.sender = sender;
    }

    fn upload(&self, contract: &impl Uploadable) -> Result<Self::Response, CwEnvError> {
        let code_id = self.app.borrow_mut().store_code(contract.wrapper());
        // add contract code_id to events manually
        let mut event = Event::new("store_code");
        event = event.add_attribute("code_id", code_id.to_string());
        let resp = AppResponse {
            events: vec![event],
            ..Default::default()
        };
        Ok(resp)
    }

    fn execute<E: Serialize + Debug>(
        &self,
        exec_msg: &E,
        coins: &[cosmwasm_std::Coin],
        contract_address: &Addr,
    ) -> Result<Self::Response, CwEnvError> {
        self.app
            .borrow_mut()
            .execute_contract(
                self.sender.clone(),
                contract_address.to_owned(),
                exec_msg,
                coins,
            )
            .map_err(From::from)
    }

    fn instantiate<I: Serialize + Debug>(
        &self,
        code_id: u64,
        init_msg: &I,
        label: Option<&str>,
        admin: Option<&Addr>,
        coins: &[cosmwasm_std::Coin],
    ) -> Result<Self::Response, CwEnvError> {
        let addr = self.app.borrow_mut().instantiate_contract(
            code_id,
            self.sender.clone(),
            init_msg,
            coins,
            label.unwrap_or("contract_init"),
            admin.map(|a| a.to_string()),
        )?;
        // add contract address to events manually
        let mut event = Event::new("instantiate");
        event = event.add_attribute("_contract_address", addr);
        let resp = AppResponse {
            events: vec![event],
            ..Default::default()
        };
        Ok(resp)
    }

    fn migrate<M: Serialize + Debug>(
        &self,
        migrate_msg: &M,
        new_code_id: u64,
        contract_address: &Addr,
    ) -> Result<Self::Response, CwEnvError> {
        self.app
            .borrow_mut()
            .migrate_contract(
                self.sender.clone(),
                contract_address.clone(),
                migrate_msg,
                new_code_id,
            )
            .map_err(From::from)
    }
}

impl<S: StateInterface> BankSetter for Mock<S> {
    fn set_balance(
        &mut self,
        address: impl Into<String>,
        amount: Vec<Coin>,
    ) -> Result<(), <Self as TxHandler>::Error> {
        (*self).set_balance(address, amount)
    }
}

#[cfg(test)]
mod test {

    use cosmwasm_std::{
        coins, to_json_binary, Addr, Binary, Coin, Deps, DepsMut, Env, MessageInfo, Response,
        StdResult, Uint128,
    };
    use cw_multi_test::ContractWrapper;
    use cw_orch_core::environment::QueryHandler;
    use serde::Serialize;
    use speculoos::prelude::*;

    use crate::core::*;

    const SENDER: &str = "cosmos123";
    const BALANCE_ADDR: &str = "cosmos456";

    #[derive(Debug, Serialize)]
    struct MigrateMsg {}

    fn execute(
        _deps: DepsMut,
        _env: Env,
        _info: MessageInfo,
        msg: cw20::Cw20ExecuteMsg,
    ) -> Result<Response, cw20_base::ContractError> {
        match msg {
            cw20::Cw20ExecuteMsg::Mint { recipient, amount } => Ok(Response::default()
                .add_attribute("action", "mint")
                .add_attribute("recipient", recipient)
                .add_attribute("amount", amount)),
            _ => unimplemented!(),
        }
    }

    fn query(_deps: Deps, _env: Env, msg: cw20_base::msg::QueryMsg) -> StdResult<Binary> {
        match msg {
            cw20_base::msg::QueryMsg::Balance { address } => Ok(to_json_binary::<Response>(
                &Response::default()
                    .add_attribute("address", address)
                    .add_attribute("balance", String::from("0")),
            )
            .unwrap()),
            _ => unimplemented!(),
        }
    }

    #[test]
    fn mock() {
        let recipient = BALANCE_ADDR;
        let sender = SENDER;
        let chain = Mock::new(sender);
        let amount = 1000000u128;
        let denom = "uosmo";

        chain
            .set_balance(recipient, vec![Coin::new(amount, denom)])
            .unwrap();
        let balance = chain.query_balance(recipient, denom).unwrap();

        asserting("address balance amount is correct")
            .that(&amount)
            .is_equal_to(balance.u128());

        asserting("sender is correct")
            .that(&sender.to_string())
            .is_equal_to(chain.sender().to_string());

        let contract_source = Box::new(
            ContractWrapper::new(execute, cw20_base::contract::instantiate, query)
                .with_migrate(cw20_base::contract::migrate),
        );

        let init_res = chain.upload_custom("cw20", contract_source).unwrap();
        asserting("contract initialized properly")
            .that(&init_res.events[0].attributes[0].value)
            .is_equal_to(&String::from("1"));

        let init_msg = cw20_base::msg::InstantiateMsg {
            name: String::from("Token"),
            symbol: String::from("TOK"),
            decimals: 6u8,
            initial_balances: vec![],
            mint: None,
            marketing: None,
        };
        let init_res = chain
            .instantiate(1, &init_msg, None, Some(&Addr::unchecked(sender)), &[])
            .unwrap();

        let contract_address = Addr::unchecked(&init_res.events[0].attributes[0].value);

        let exec_res = chain
            .execute(
                &cw20_base::msg::ExecuteMsg::Mint {
                    recipient: recipient.to_string(),
                    amount: Uint128::from(100u128),
                },
                &[],
                &contract_address,
            )
            .unwrap();

        asserting("that exect passed on correctly")
            .that(&exec_res.events[1].attributes[1].value)
            .is_equal_to(&String::from("mint"));

        let query_res = chain
            .query::<cw20_base::msg::QueryMsg, Response>(
                &cw20_base::msg::QueryMsg::Balance {
                    address: recipient.to_string(),
                },
                &contract_address,
            )
            .unwrap();

        asserting("that query passed on correctly")
            .that(&query_res.attributes[1].value)
            .is_equal_to(&String::from("0"));

        let migration_res = chain.migrate(&cw20_base::msg::MigrateMsg {}, 1, &contract_address);
        asserting("that migration passed on correctly")
            .that(&migration_res)
            .is_ok();
    }

    #[test]
    fn custom_mock_env() {
        let mock_state = MockState::new();
        let chain = Mock::<_>::new_custom(SENDER, mock_state);

        let recipient = BALANCE_ADDR;
        let amount = 1000000u128;
        let denom = "uosmo";

        chain
            .set_balances(&[(recipient, &[Coin::new(amount, denom)])])
            .unwrap();

        let balances = chain.query_all_balances(recipient).unwrap();
        asserting("recipient balances length is 1")
            .that(&balances.len())
            .is_equal_to(1);
    }

    #[test]
    fn state_interface() {
        let contract_id = "my_contract";
        let code_id = 1u64;
        let address = &Addr::unchecked(BALANCE_ADDR);
        let mut mock_state = Rc::new(RefCell::new(MockState::new()));

        mock_state.set_address(contract_id, address);
        asserting!("that address has been set")
            .that(&address)
            .is_equal_to(&mock_state.get_address(contract_id).unwrap());

        mock_state.set_code_id(contract_id, code_id);
        asserting!("that code_id has been set")
            .that(&code_id)
            .is_equal_to(mock_state.get_code_id(contract_id).unwrap());

        asserting!("that total code_ids is 1")
            .that(&mock_state.get_all_code_ids().unwrap().len())
            .is_equal_to(1);

        asserting!("that total addresses is 1")
            .that(&mock_state.get_all_addresses().unwrap().len())
            .is_equal_to(1);
    }

    #[test]
    fn add_balance() {
        let chain = Mock::new(SENDER);
        let recipient = BALANCE_ADDR;
        let amount = 1000000u128;
        let denom_1 = "uosmo";
        let denom_2 = "osmou";

        chain
            .add_balance(recipient, vec![Coin::new(amount, denom_1)])
            .unwrap();
        chain
            .add_balance(recipient, vec![Coin::new(amount, denom_2)])
            .unwrap();

        let balances = chain.query_all_balances(recipient).unwrap();
        asserting("recipient balances added")
            .that(&balances)
            .contains_all_of(&[&Coin::new(amount, denom_1), &Coin::new(amount, denom_2)])
    }

    #[test]
    fn bank_querier_works() -> Result<(), CwEnvError> {
        let denom = "urandom";
        let init_coins = coins(45, denom);
        let sender = "sender";
        let app = Mock::new(sender);
        app.set_balance(sender, init_coins.clone())?;
        let sender = app.sender.clone();
        assert_eq!(
            app.bank_querier()
                .balance(sender.clone(), Some(denom.to_string()))?,
            init_coins
        );
        assert_eq!(
            app.bank_querier().supply_of(denom.to_string())?,
            init_coins[0]
        );

        Ok(())
    }
}
