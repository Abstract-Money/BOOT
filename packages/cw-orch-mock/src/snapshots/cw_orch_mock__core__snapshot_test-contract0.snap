---
source: packages/cw-orch-mock/src/core.rs
expression: self.app.borrow().dump_wasm_raw(&contract_addr)
---
- - - 99
    - 111
    - 110
    - 116
    - 114
    - 97
    - 99
    - 116
    - 95
    - 105
    - 110
    - 102
    - 111
  - - 123
    - 34
    - 99
    - 111
    - 110
    - 116
    - 114
    - 97
    - 99
    - 116
    - 34
    - 58
    - 34
    - 99
    - 114
    - 97
    - 116
    - 101
    - 115
    - 46
    - 105
    - 111
    - 58
    - 99
    - 111
    - 117
    - 110
    - 116
    - 101
    - 114
    - 34
    - 44
    - 34
    - 118
    - 101
    - 114
    - 115
    - 105
    - 111
    - 110
    - 34
    - 58
    - 34
    - 48
    - 46
    - 49
    - 49
    - 46
    - 48
    - 34
    - 125
- - - 115
    - 116
    - 97
    - 116
    - 101
  - - 123
    - 34
    - 99
    - 111
    - 117
    - 110
    - 116
    - 34
    - 58
    - 49
    - 44
    - 34
    - 111
    - 119
    - 110
    - 101
    - 114
    - 34
    - 58
    - 34
    - 115
    - 101
    - 110
    - 100
    - 101
    - 114
    - 34
    - 125

